File: myExpenses/src/main/java/eltos/simpledialogfragment/form/AmountInputViewHolder.java
package eltos.simpledialogfragment.form;

import android.content.Context;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;

import com.google.android.material.textfield.TextInputLayout;

import org.totschnig.myexpenses.R;

import java.math.BigDecimal;

import butterknife.BindView;
import butterknife.ButterKnife;

class AmountInputViewHolder extends FormElementViewHolder<AmountInput> {
  @BindView(R.id.inputLayout)
  TextInputLayout inputLayout;
  @BindView(R.id.amount)
  org.totschnig.myexpenses.ui.AmountInput amountInputText;

  protected AmountInputViewHolder(AmountInput field) {
    super(field);
  }

  @Override
  protected int getContentViewLayout() {
    return R.layout.simpledialogfragment_form_item_amount;
  }

  @Override
  protected void setUpView(View view, Context context, Bundle savedInstanceState, SimpleFormDialog.DialogActions actions) {
    ButterKnife.bind(this, view);
    inputLayout.setHint(field.getText(context));
    amountInputText.setFractionDigits(field.fractionDigits);
    if (field.withTypeSwitch == null) {
      amountInputText.setWithTypeSwitch(false);
    } else {
      amountInputText.setWithTypeSwitch(true);
      amountInputText.setType(field.withTypeSwitch);
    }
    if (field.amount != null) {
      amountInputText.setAmount(field.amount);
    }
    // Positive button state for single element forms
    if (actions.isOnlyFocusableElement()) {
      amountInputText.addTextChangedListener(new TextWatcher() {
        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {
        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {
        }

        @Override
        public void afterTextChanged(Editable s) {
          actions.updatePosButtonState();
        }
      });
    }
  }

  @Override
  protected void saveState(Bundle outState) {

  }

  @Override
  protected void putResults(Bundle results, String key) {
    results.putSerializable(key, amountInputText.getTypedValue());
  }

  @Override
  protected boolean focus(SimpleFormDialog.FocusActions actions) {
    return amountInputText.requestFocus();
  }

  @Override
  protected boolean posButtonEnabled(Context context) {
    if (!field.required) return true;
    return amountInputText.validate(false) != null;
  }

  @Override
  protected boolean validate(Context context) {
    final BigDecimal result = amountInputText.validate(true);
    if (result == null) return false;
    if (field.max != null && result.compareTo(field.max) > 0) {
      inputLayout.setError(field.maxExceededError);
      return false;
    }
    if (field.min != null && result.compareTo(field.min) < 0) {
      inputLayout.setError(field.underMinError);
      return false;
    }
    return true;
  }
}

File: myExpenses/src/main/java/eltos/simpledialogfragment/form/AmountInputViewHolder.kt
package eltos.simpledialogfragment.form

import android.content.Context
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import com.google.android.material.textfield.TextInputLayout
import eltos.simpledialogfragment.form.SimpleFormDialog.DialogActions
import eltos.simpledialogfragment.form.SimpleFormDialog.FocusActions
import org.totschnig.myexpenses.R
import org.totschnig.myexpenses.model.Money.Companion.convertBigDecimal
import org.totschnig.myexpenses.ui.MyTextWatcher

internal class AmountInputViewHolder(field: AmountInput) : FormElementViewHolder<AmountInput>(field) {
    private lateinit var inputLayout: TextInputLayout
    private lateinit var amountInput: org.totschnig.myexpenses.ui.AmountInput
    override fun getContentViewLayout(): Int {
        return R.layout.simpledialogfragment_form_item_amount
    }

    override fun setUpView(
        view: View,
        context: Context,
        savedInstanceState: Bundle?,
        actions: DialogActions
    ) {
        inputLayout = view.findViewById(R.id.inputLayout)
        amountInput = view.findViewById(R.id.amount)
        inputLayout.hint = field.getText(context)
        amountInput.setFractionDigits(field.fractionDigits)
        field.getText(context)?.let {
            amountInput.contentDescription = it
        }
        field.withTypeSwitch?.also {
            amountInput.setWithTypeSwitch(true)
            amountInput.type = it
        } ?: run {
            amountInput.setWithTypeSwitch(false)
        }
        field.amount?.let {
            amountInput.setAmount(it)
        }
        // Positive button state for single element forms
        if (actions.isOnlyFocusableElement) {
            amountInput.addTextChangedListener(object : MyTextWatcher() {
                override fun afterTextChanged(s: Editable) {
                    actions.updatePosButtonState()
                }
            })
        }
    }

    override fun saveState(outState: Bundle) {}
    override fun putResults(results: Bundle, key: String) {
        results.putSerializable(key, amountInput.typedValue)
    }

    override fun focus(actions: FocusActions): Boolean {
        return amountInput.requestFocus()
    }

    override fun posButtonEnabled(context: Context): Boolean {
        return if (!field.required) true else amountInput.getUntypedValue(false).getOrNull() != null
    }

    override fun validate(context: Context): Boolean {
        val result = amountInput.getUntypedValue(true).getOrNull() ?: return false
        try {
            convertBigDecimal(result, field.fractionDigits)
        } catch (e: ArithmeticException) {
            inputLayout.setError("Number too large")
            return false
        }
        if (field.max != null && result > field.max) {
            inputLayout.setError(field.maxExceededError)
            return false
        }
        if (field.min != null && result < field.min) {
            inputLayout.setError(field.underMinError)
            return false
        }
        return true
    }
}
